apt/utils.py

```diff
@@ -21,25 +24,68 @@ def stream_file_transfer(
     Args:
         source_uri: Source URI or path to read from
         target_uri: Target URI or path to write to
-        chunk_size: Size of chunks to read/write in bytes
+        max_concurrency: Maximum number of concurrent threads for S3 transfers
+        chunk_size: Size of chunks for transfers in bytes
     """
     source_uri = str(source_uri)
     target_uri = str(target_uri)
 
     logger.debug(f"Streaming file from '{source_uri}' to '{target_uri}'")
 
-    # ensure parent directory exists for local target paths
-    if not target_uri.startswith("s3://"):
-        target_path = Path(target_uri)
-        target_path.parent.mkdir(parents=True, exist_ok=True)
+    s3_client = boto3.client("s3")
+    transfer_config = TransferConfig(
+        multipart_threshold=chunk_size,
+        max_concurrency=max_concurrency,
+        multipart_chunksize=chunk_size,
+        use_threads=True,
+    )
 
-    total_bytes = 0
-    with open(source_uri, "rb") as source_file, open(target_uri, "wb") as target_file:
-        while True:
-            chunk = source_file.read(chunk_size)
-            if not chunk:
-                break
-            target_file.write(chunk)
-            total_bytes += len(chunk)
+    # S3-to-Local
```

Comment:
Commenting here, but applicable to all code blocks.

I had considered dedicated function like `_s3_to_local_transfer()` and `_local_to_s3_transfer()`, etc., but it felt like overkill.